macrosynergy/download/dq_api.py

```diff
@@ -0,0 +1,629 @@
+"""DataQuery Interface."""
+import warnings
+import concurrent.futures
+import time
+import datetime
+import logging
+from math import ceil, floor
+from itertools import chain
+
+import base64
+import os
+import requests
+from typing import List, Optional, Dict, Tuple
+from datetime import datetime
+
+CERT_BASE_URL: str = "https://platform.jpmorgan.com/research/dataquery/api/v2"
+OAUTH_BASE_URL: str = (
+    "https://api-developer.jpmorgan.com/research/dataquery-authe/api/v2"
+)
+OAUTH_TOKEN_URL: str = "https://authe.jpmchase.com/as/token.oauth2"
+OAUTH_DQ_RESOURCE_ID: str = "JPMC:URI:RS-06785-DataQueryExternalApi-PROD"
+
+logger = logging.getLogger(__name__)
+
+
+def valid_response(r: requests.Response) -> Tuple[dict, bool, Optional[dict]]:
+    """
+    Prior to requesting any data, the function will confirm if a connection to the
+    DataQuery API is able to be established given the credentials passed. If the status
+    code is 200, able to access DataQuery's API.
+    """
+    msg: Optional[dict] = None
+    if not r.ok:
+        msg: Dict[str, str] = {
+            "headers": r.headers,
+            "status_code": r.status_code,
+            "text": r.text,
+            "url": r.url,
+        }
+        js: Optional[dict] = None
+    else:
+        js = r.json()
+
+    return js, r.ok, msg
+
+
+def dq_request(
+    url: str,
+    headers: dict = None,
+    params: dict = None,
+    method: str = "get",
+    cert: Optional[Tuple[str, str]] = None,
+    **kwargs,
+) -> Tuple[dict, str, str]:
+    """Will return the request from DataQuery."""
+    request_error = f"Unknown request method {method} not in ('get', 'post')."
+    assert method in ("get", "post"), request_error
+
+    with requests.request(
+        method=method,
+        url=url,
+        cert=cert,
+        headers=headers,
+        params=params,
+        **kwargs,
+    ) as r:
+        text: str = r.text
```

Comment:
It looks to me like we don't need the `text` variable?

Also, isn't `last_url` captured in `msg` (or we can add it to always be captured in msg?